external-crates/move/crates/move-compiler/src/typing/deprecation_warnings.rs

```diff
@@ -0,0 +1,252 @@
+// Copyright (c) The Move Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    diag,
+    expansion::ast::{self as E, ModuleIdent},
+    ice,
+    shared::{
+        known_attributes::{AttributePosition, DeprecationAttribute, KnownAttribute},
+        program_info::NamingProgramInfo,
+        CompilationEnv, Name,
+    },
+};
+use move_ir_types::location::Loc;
+use std::collections::HashMap;
+
+const NOTE_STR: &str = "note";
+
+#[derive(Debug, Clone)]
+pub struct Deprecation {
+    // The source location of the deprecation attribute
+    pub source_location: Loc,
+    // The type of the member that is deprecated (function, constant, etc.)
+    pub location: AttributePosition,
+    // The module that the deprecated member belongs to. This is used in part to make sure we don't
+    // register deprecation warnings for members within a given deprecated module calling within
+    // that module.
+    pub module_ident: ModuleIdent,
+    // Information about the deprecation information depending on the deprecation attribute.
+    // #[deprecated]  -- if None
+    // #[deprecated(note = b"message")] -- if Some(message)
+    pub deprecation_note: Option<String>,
+}
+
+#[derive(Debug, Clone)]
+pub struct Deprecations {
+    // Name = None -- deprecation on Module
+    // Name = Some(Name) -- deprecation on Module::Name member
+    pub deprecated_members: HashMap<(ModuleIdent, Option<Name>), Deprecation>,
+}
+
+impl Deprecations {
+    /// Index the modules and their members for deprecation attributes and register each
+    /// deprecation attribute for use later on.
+    pub fn index_deprecations(env: &mut CompilationEnv, info: &NamingProgramInfo) -> Self {
```

Comment:
Fine with this but maybe also 
```suggestion
    pub fn new(env: &mut CompilationEnv, info: &NamingProgramInfo) -> Self {
```